### ISSUE TITLE
Stop supporting .war deployments. Support standalone applications!

### ISSUE THREAD

**Issue by tkrille** [osiam/osiam#5]
Ok, let's face it: application servers are dead. But we still _support_ (I really mean **depend on**) them for nothing at all. We want to move to Spring Boot to support standalone applications, improved configuration and easier development, testing, deployment, etc. But it seems rather hard to support .war deployments too. @dacrome already did some incredible work on this topic and I think we should start to discuss this more openly and transparently.

IIRC the first candidate is `addon-self-administration` as it is the most easiest to migrate regarding the Spring configuration stuff. It also has just a small impact as it is a supporting component that can be easily replaced. I think most work is already done by @dacrome, so it's only a question of "when will we do it?".

Here comes the list with links to the issues in the affected repos. It's intended for checking each completed item.
- [ ] #69
- [ ] osiam/addon-self-administration#133
- [ ] osiam/addon-administration#88

**Comment by dacrome** [comment]
+:100: 
I think I will create a PR of the Spring Boot migration later this week of the `addon-self-administration`. To discuss some ideas and how to handle configuration and stuff, I will create issues in the `addon-self-administration` [repo](https://github.com/osiam/addon-self-administration/issues).
We should think about a migration documentation to provide an easy switch for operations.
The collected know-how of the migration to Spring Boot should be used to also move `OSIAM`, which is already in the making.

**Comment by dacrome** [comment]
The steps to migrate all components to Spring Boot fat jar should be the following:
1. Switch to java configuration (maybe except the [auth-server](https://github.com/osiam/auth-server/issues/23#issuecomment-110486934), addons already uses only java config)
2. Migrate to Spring Boot fat war, configuration like before, deployable in tomcat and also run with java -jar (addons are migrated)
3. Migrate to Spring Boot fat jar with advanced external configuration (e.g, lime discussed in [self-admin](https://github.com/osiam/addon-self-administration/issues/122))

**Comment by timopick** [comment]
Full ack, who needs Tomcat?! One small objection, tho. I wonder if it's always preferable to have java config over xml. After all, the java config needs a recompile to make changes - xml on the other hand can be edited in-place. This can really be handy for defining custom scopes / voters. Am I wrong here?

**Comment by dacrome** [comment]
I don't think that the xml configuration is reloadable, maybe with some black magic or JRebel and only during development. I think in production this shouldn't be done, even if it's possible. Just think of singleton scoped beans which are used application wide.
The scopes are predefined and aren't changeable at runtime as well as the voters. Dynamic scopes can only be achieved by combining scopes. That's why OSIAM should provide as many useful scopes as possible. Any other complex authorization have to be done by the client application.

BTW: Spring Boot 1.3.0 comes with some fancy stuff to [improve the experience during development](https://github.com/spring-projects/spring-boot/issues/3082).

**Comment by tpick** [comment]
I know that restarting the app is mandatory for either way (xml or java config). But what I meant was, that it is not possible to make changes to the java config without recompilation of the source code, right?

**Comment by dacrome** [comment]
Ah sorry, I misunderstood your comment ... But: any dynamic configuration should be done by the [configuration files](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config-application-property-files) whose changes are available after restart, but the scopes should be predefined anyway...

**Comment by tpick** [comment]
What about custom scopes? I can think of scenarios where a user wants to add scopes that are not provided in the default installation. Wouldn't it be easier for them to tailor their osiam installation if they had full control over the bean declaration via xml?

**Comment by tkrille** [comment]
Scopes are added for a Client directly in the database. So you can always
introduce new scopes. On the other hand, there must be a component that
actually evaluates the given scopes and makes an authorization decision.
This second part is hard to make dynamic in the core components of OSIAM
and it also seems a little bit like YAGNI. Nevertheless, if there is a real
use case for this, we'll find a proper way. But you can still do this in
your own application that uses OSIAM, e.g. you can define a scope
`VIEW_ALL_ORDERS` and use that in your shop application to give the order
processing department's members the right to view all orders.

BTW: It's planned that you can configure extra scopes in the configuration
of OSIAM, so you don't need to hack the database to introduce a
new scope. There's also the idea to create a RESTful resource for scope
management and define custom clients in the OSIAM's configuration,
too.

**Comment by tpick** [comment]
Alrighty. Thanks for the explanation!

**Comment by tkrille** [comment]
I've updated the initial comment with a checklist to show the current status of this undertaking and to link to the corresponding umbrella tasks of the affected components.

**Comment by tkrille** [comment]
Alright, we've reached the first milestone some time ago, namely migrating all services to Spring Boot. So, the next milestone is actually producing standalone fat JARs. But before we can go on, some things have to be clarified: how to structure the YAML configuration file and how to ship and deploy static assets (OSIAM and addon-self-administration). I'll create 2 separate issues for this, so that we can discuss these topics.

**Comment by tkrille** [comment]
YAML config: https://github.com/osiam/osiam/issues/35

**Comment by tkrille** [comment]
Assets: https://github.com/osiam/osiam/issues/36

**Comment by tkrille** [comment]
I think the next steps are also still blocked by https://github.com/osiam/osiam/issues/6.

**Comment by tkrille** [comment]
I think we can still support WARs. After all it's already possible to put the WARs in a servlet container and run it on the command line via `java -cp osiam.war:/etc/osiam org.springframework.boot.loader.WarLauncher`.

**Comment by jvmlet** [comment]
osiam-spring-boot-starter would be great feature.

**Comment by tkrille** [comment]
@jvmlet: To embed OSIAM in your own Spring Boot application?

**Comment by jvmlet** [comment]
Exactly

**Comment by tkrille** [comment]
We've already started to think about this. Might be also a viable approach for us to create a modular version of OSIAM, especially when considering the add-ons. But I don't think, that anyone has started some research or further investigations yet. If you really want to see this feature in OSIAM, please open an issue, so we can track it and plan appropriately.

**Comment by tkrille** [comment]
As we give up the add-ons (#158) this issue will be closed. Further discussion will be done in #69.
